autode/opt/optimisers/trm.py

```diff
@@ -0,0 +1,461 @@
+            step = -inv_shifted_h @ self._coords.g
+            size = np.linalg.norm(step)
+            deriv = -np.linalg.multi_dot((step, inv_shifted_h, step))
+            deriv = float(deriv) / size
+            return size, deriv, step
+
+        def optimise_lambda_for_int_step(int_size):
+            """Given a step length in internal coords, get the
+            value of lambda that will give that step"""
+            # from pyberny and geomeTRIC
+            d = 1.0  # use bisection to ensure lambda < first_b
+            for _ in range(20):
+                size, _, _ = get_int_step_size_and_deriv(first_b - d)
+                # find f(x) > 0, so going downhill has no risk of jumping over
+                if size > int_size:
+                    break
+                d = d / 2.0
+            else:
+                raise OptimiserStepError("Failed to find f(x) > 0")
+            lmda_guess = first_b - d
+            for _ in range(50):
+                size, der, _ = get_int_step_size_and_deriv(lmda_guess)
+                if abs(size - int_size) / int_size < 0.001:  # 0.1% error
+                    break
+                lmda_guess -= (1 - size / int_size) * (size / der)
+            else:
+                raise OptimiserStepError("Failed in optimising internal step")
+            return lmda_guess
+
+        last_lmda = 0.0  # initialize non-local var
+
+        def cart_step_length_error(int_size):
+            """Deviation for trust radius given step-size
+            in internal coordinates"""
+            nonlocal last_lmda
+            last_lmda = optimise_lambda_for_int_step(int_size)
+            _, _, step = get_int_step_size_and_deriv(last_lmda)
+            step_size = self._get_cart_step_size_from_step(step)
+            return step_size - self.alpha
+
+        # The value of shift parameter lambda must lie within (-infinity, first_b)
+        # Need to find the roots of the 1D function cart_step_length_error
+        int_step_size = self.alpha  # starting guess, so ok to use cartesian
+        fac = 1.0
+        size_min_bound = None
+        size_max_bound = None
+
+        for _ in range(10):
+            int_step_size = int_step_size * fac
+            err = cart_step_length_error(int_step_size)
+
+            if err < -1.0e-6:  # found where error is < 0
+                fac = 2.0  # increase the step size
+                size_min_bound = int_step_size
+
+            elif err > 1.0e-6:  # found where error is > 0
+                fac = 0.5  # decrease the step size
+                size_max_bound = int_step_size
+
+            else:  # found err ~ 0 already!, no need for root finding
+                return self._coords.h - last_lmda * np.eye(h_n)
+
+            if (size_max_bound is not None) and (size_min_bound is not None):
+                break
+        else:
+            raise OptimiserStepError(
+                "Unable to find bracket range for root finding"
+            )
```

Comment:
@t-young31 Wait why is for-else bad? I actually learned about for--else while trying to write this function (because I could not write it clean in any other way) 😓

Reply:
just because it's hinders readability (for-else is a pretty niche feature). glancing at it an `size_max_bound is None) and (size_min_bound is None):` might suffice?